File: Play/kafka/python-examples/consumer.py
#!/usr/bin/env python3
"""
Simple Kafka Consumer Example
Consumes messages from a Kafka topic
"""

from confluent_kafka import Consumer, KafkaError
import json
import sys

# Kafka configuration
config = {
    'bootstrap.servers': 'localhost:9092',
    'group.id': 'python-consumer-group',
    'auto.offset.reset': 'earliest',
    'enable.auto.commit': True
}

# Create Consumer instance
consumer = Consumer(config)


def consume_messages(topic='test-topic'):
    """Consume messages from Kafka topic"""
    print(f"📥 Starting to consume messages from topic '{topic}'...")
    print("Press Ctrl+C to stop\n")

    # Subscribe to topic
    consumer.subscribe([topic])

    try:
        message_count = 0
        while True:
            # Poll for messages
            msg = consumer.poll(timeout=1.0)

            if msg is None:
                continue

            if msg.error():
                if msg.error().code() == KafkaError._PARTITION_EOF:
                    print(f'📌 Reached end of partition {msg.partition()}')
                else:
                    print(f'❌ Error: {msg.error()}')
                continue

            # Process message
            message_count += 1
            key = msg.key().decode('utf-8') if msg.key() else None
            value = json.loads(msg.value().decode('utf-8'))

            print(f"📨 Message #{message_count}")
            print(f"   Topic: {msg.topic()}")
            print(f"   Partition: {msg.partition()}")
            print(f"   Offset: {msg.offset()}")
            print(f"   Key: {key}")
            print(f"   Value: {value}")
            print()

    except KeyboardInterrupt:
        print("\n⚠️  Consumer interrupted by user")
    except Exception as e:
        print(f"❌ Error: {e}")
    finally:
        # Close consumer
        print("🔒 Closing consumer...")
        consumer.close()


if __name__ == '__main__':
    topic = sys.argv[1] if len(sys.argv) > 1 else 'test-topic'
    consume_messages(topic)

File: Play/kafka/go-examples/consumer.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"os/signal"
	"syscall"

	"github.com/IBM/sarama"
)

// Message represents the data structure for Kafka messages
type Message struct {
	ID        int    `json:"id"`
	Message   string `json:"message"`
	Timestamp int64  `json:"timestamp"`
}

// ConsumerGroupHandler implements sarama.ConsumerGroupHandler
type ConsumerGroupHandler struct {
	messageCount int
}

func (h *ConsumerGroupHandler) Setup(sarama.ConsumerGroupSession) error {
	return nil
}

func (h *ConsumerGroupHandler) Cleanup(sarama.ConsumerGroupSession) error {
	return nil
}

func (h *ConsumerGroupHandler) ConsumeClaim(session sarama.ConsumerGroupSession, claim sarama.ConsumerGroupClaim) error {
	for message := range claim.Messages() {
		h.messageCount++

		var msg Message
		if err := json.Unmarshal(message.Value, &msg); err != nil {
			log.Printf("❌ Error unmarshaling message: %v", err)
			continue
		}

		fmt.Printf("📨 Message #%d\n", h.messageCount)
		fmt.Printf("   Topic: %s\n", message.Topic)
		fmt.Printf("   Partition: %d\n", message.Partition)
		fmt.Printf("   Offset: %d\n", message.Offset)
		fmt.Printf("   Key: %s\n", string(message.Key))
		fmt.Printf("   Value: %+v\n\n", msg)

		session.MarkMessage(message, "")
	}

	return nil
}

func main() {
	// Get parameters
	topic := "test-topic"
	if len(os.Args) > 1 {
		topic = os.Args[1]
	}

	// Kafka configuration
	config := sarama.NewConfig()
	config.Version = sarama.V2_6_0_0
	config.Consumer.Group.Rebalance.Strategy = sarama.NewBalanceStrategyRoundRobin()
	config.Consumer.Offsets.Initial = sarama.OffsetOldest

	// Create consumer group
	consumerGroup, err := sarama.NewConsumerGroup([]string{"localhost:9092"}, "go-consumer-group", config)
	if err != nil {
		log.Fatalf("❌ Failed to create consumer group: %v", err)
	}
	defer consumerGroup.Close()

	fmt.Printf("📥 Starting to consume messages from topic '%s'...\n", topic)
	fmt.Println("Press Ctrl+C to stop\n")

	// Handle signals
	ctx, cancel := context.WithCancel(context.Background())
	sigterm := make(chan os.Signal, 1)
	signal.Notify(sigterm, syscall.SIGINT, syscall.SIGTERM)

	handler := &ConsumerGroupHandler{}

	// Consumer loop
	go func() {
		for {
			if err := consumerGroup.Consume(ctx, []string{topic}, handler); err != nil {
				log.Printf("❌ Error from consumer: %v", err)
			}

			if ctx.Err() != nil {
				return
			}
		}
	}()

	<-sigterm
	fmt.Println("\n⚠️  Consumer interrupted by user")
	cancel()
	fmt.Println("🔒 Closing consumer...")
}
